code/game/machinery/telecomms/broadcasting.dm

```diff
@@ -53,127 +53,141 @@
+					radios += R
+
+		if (TRANSMISSION_SUPERSPACE)
+			// Only radios which are independent
+			for(var/obj/item/device/radio/R in GLOB.all_radios["[frequency]"])
+				if(R.independent && R.can_receive(frequency, levels))
+					radios += R
+
+	// From the list of radios, find all mobs who can hear those.
+	var/list/receive = get_mobs_in_radio_ranges(radios)
+
+	// Cut out mobs with clients who are admins and have radio chatter disabled.
+	for(var/mob/R in receive)
+		if (R.client && R.client.holder && !(R.client.prefs.chat_toggles & CHAT_RADIO))
 			receive -= R
 
-	for(var/mob/M in GLOB.player_list)
-		if(isobserver(M) && M.client && (M.client.prefs.chat_toggles & CHAT_GHOSTRADIO))
+	// Add observers who have ghost radio enabled.
+	for(var/mob/dead/observer/M in GLOB.player_list)
+		if(M.client && (M.client.prefs.chat_toggles & CHAT_GHOSTRADIO))
 			receive |= M
 
-	var/rendered = virt.compose_message(virt, language, message, freq, spans) //Always call this on the virtualspeaker to advoid issues.
+	// Render the message and have everybody hear it.
+	// Always call this on the virtualspeaker to avoid issues.
+	var/spans = data["spans"]
+	var/rendered = virt.compose_message(virt, language, message, frequency, spans)
 	for(var/atom/movable/hearer in receive)
-		hearer.Hear(rendered, virt, language, message, freq, spans)
+		hearer.Hear(rendered, virt, language, message, frequency, spans)
 
+	// This following recording is intended for research and feedback in the use of department radio channels
 	if(length(receive))
-		// --- This following recording is intended for research and feedback in the use of department radio channels ---
-		SSblackbox.LogBroadcast(freq)
-
-	sleep(50)
-	if(!QDELETED(virt)) //It could happen to YOU
-		qdel(virt)
-
-//Use this to test if an obj can communicate with a Telecommunications Network
-
-/atom/proc/telecomms_process()
-
-	// First, we want to generate a new radio signal
-	var/datum/signal/signal = new
-	signal.transmission_method = TRANSMISSION_SUBSPACE
-	var/turf/pos = get_turf(src)
-
-	// --- Finally, tag the actual signal with the appropriate values ---
-	signal.data = list(
-		"slow" = 0, // how much to sleep() before broadcasting - simulates net lag
-		"message" = "TEST",
-		"compression" = rand(45, 50), // If the signal is compressed, compress our message too.
-		"traffic" = 0, // dictates the total traffic sum that the signal went through
-		"type" = 4, // determines what type of radio input it is: test broadcast
-		"reject" = 0,
-		"done" = 0,
-		"level" = pos.z // The level it is being broadcasted at.
-	)
-	signal.frequency = FREQ_COMMON
-
-  //#### Sending the signal to all subspace receivers ####//
-	for(var/obj/machinery/telecomms/receiver/R in GLOB.telecomms_list)
-		R.receive_signal(signal)
-
-	sleep(rand(10,25))
+		SSblackbox.LogBroadcast(frequency)
 
-	return signal
+	QDEL_IN(virt, 50)  // Make extra sure the virtualspeaker gets qdeleted
```

Comment:
Do signals actually stay around after use for this to fire ?

Reply:
The signal doesn't need to stay around, `QDEL_IN` sets a global timer:
```dm
#define QDEL_IN(item, time) addtimer(CALLBACK(GLOBAL_PROC, .proc/qdel, item), time, TIMER_STOPPABLE)
```